Join stake pool shelley

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#1709 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have mocked stake pool layer and added joinStakePool to shelley api server
- [x] I have added first tests (now wallet id and no pool id) for stake pools in shelley
- [x] I have added impl for mkDelegationJoinTx
- [x] I have extended mkDelegationJoinTx by adding WalletDelegation 
- [x] I have added tests for joining/rejoining
- [x] I have impl quitting and added tests

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
add joinStakePool to Server and mock StakePoolLayer

paweljakubas committed:
add StakePools integration suite

paweljakubas committed:
mkDelegationJoinTx - first version

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -134,6 +148,56 @@ newTransactionLayer _proxy _protocolMagic epochLength = TransactionLayer
+            (k, pwd) <- lookupPrivateKey keyFrom addr
+            pure $ mkWitness unsigned (getRawKey k, pwd)
+
+        witsForChimeric <- fmap Set.fromList (pure $ [mkWitness unsigned (getRawKey accXPrv, pwd')])
```
this is probably the culprit for too low fee - I am adding this artificially

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -134,6 +147,62 @@ newTransactionLayer _proxy _protocolMagic epochLength = TransactionLayer
+        -> [TxOut]
+        -> Either ErrMkTx (Tx, SealedTx)
+    _mkDelegationJoinTx poolId (accXPrv, pwd') keyFrom slot ownedIns outs = do
+        let timeToLive = (toSlotNo epochLength slot) + 7200
```
I know this is still a draft, but I hope the similarities between `mkDelegationJoinTx` and `mkStdTx` can be factored out

paweljakubas replied:
yes, some code overlap - I will reuse it later

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -134,6 +147,62 @@ newTransactionLayer _proxy _protocolMagic epochLength = TransactionLayer
+                VerKeyEd25519DSIGN .
+                throwCryptoError $
+                Ed25519.publicKey pId
+                --}
```
I'd definitely factor these functions out to make the body of this function more readable :+1:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -134,6 +147,62 @@ newTransactionLayer _proxy _protocolMagic epochLength = TransactionLayer
+                :: PoolId
+                -> Cardano.ShelleyVerificationKeyHashStakePool
+            toPoolIdHash (PoolId pId) =
+                SL.KeyHash $ Hash.UnsafeHash pId
```
Shouldn't our pool id be already the hashed version?

paweljakubas replied:
right now it is hashed, so it goes wrapped with ledger data constructors. Below, there is version that hashes verification key. But because `knownStakePools` will return hashes we are not going to use it

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -134,6 +147,62 @@ newTransactionLayer _proxy _protocolMagic epochLength = TransactionLayer
+                (toStakingKeyHash accXPrv)
+                (toPoolIdHash poolId)
+        let unsigned =
+                mkUnsignedTx timeToLive ownedIns outs [registerCert, delegateCert]
```
:+1:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -155,14 +224,18 @@ _estimateSize (CoinSelection inps outs chngs) =
 
+        dummyKeyHash = SL.KeyHash . Hash.UnsafeHash $ BS.pack (1:replicate 64 0)
+        dummyRegisterCert = Cardano.shelleyRegisterStakingAddress dummyKeyHash
+        dummyDelegateCert = Cardano.shelleyDelegateStake dummyKeyHash dummyKeyHash
```
We should make the presence of certificate a parameter of `estimateSize` and tweak the functions of the wallet engine to tell whether there's or there isn't some certificate to acknowledge.

KtorZ committed:
extract delegation certificates from transaction body when receiving blocks

KtorZ committed:
adjust 'toChimericAccount' to return a hashed key

This also impacts Jörmungandr and _probably_ (haven't give it much
thoughts) require a migration that goes through all stored certificate
and hash their existing chimeric account.

KtorZ committed:
factor out time to live in shelley transaction layer

KtorZ committed:
move compatibility function for certificate creation in proper module

KtorZ committed:
revise fee calculation to allow distinguishing delegation tx from normal payment

This demanded some rework that also impacts Jörmungandr. Since the fee calculation itself is actually different between both backend,
it made sense to move it entirely to the transaction layer so that each backend can implement its own version of the fee calculation
on top of an accurate size estimation

KtorZ committed:
make pool operator credentials generation deterministic

So we can rely on a hard-coded list of pool ids for testing

paweljakubas committed:
change to existing pool id

KtorZ committed:
fix jormungandr delegation certificate creation

A previous change in 'toChimericAccount' indeed had an impact on the Jörmungandr transaction layer which must
not hash the reward account.

paweljakubas committed:
add rejoining stake pool test

paweljakubas committed:
extend mkDelegationJoinTx

paweljakubas committed:
correct jormungandr unit tests and show that discovery and walletDelegation changes work

paweljakubas committed:
support quitting stake pool in shelley

paweljakubas committed:
enable quit stake pool in shelley server and add test

KtorZ committed:
minor neatpicks while reviewing code.

KtorZ reviewed (approved):
:heavy_check_mark:

KtorZ committed:
remove redundant steps in last stake pool integration scenario

KtorZ committed:
reduce k parameter in genesis and increase d (more blocks by the BFT leader)

rvl committed:
Retry stake pool registrations in integration tests

The transaction submission command seems to occasionally fail on
macos, unable to open the socket.

KtorZ committed:
set default shelley cluster severity to Error

Most messages from the node are actually redundant with the messages we
get from our own networking layer.

KtorZ committed:
hopefully temporary disable running pools in the integration cluster

Pools are still registered on-chain beforehand, so we can still get to test a bunch of things.
